# Naive Bayes Classifier from the scratch

Implemented Naive Bayes Algoritm from the scratch for a small data set having only two feature.No any Library is used for Machine Learnig Algoritm.I have used only python library Panday for Data Preparation and Numpy Library for computational work.

In [2]:
import pandas as pd
import numpy as np

In [3]:
df1=pd.read_csv('ex2data1.csv')
df=df1.iloc[20:]
df2=df1.iloc[:20]
df2.head()

x          y  label
0  34.623660  78.024693      0
1  30.286711  43.894998      0
2  35.847409  72.902198      0
3  60.182599  86.308552      1
4  79.032736  75.344376      1

In [4]:
data_0=df.loc[df['label'] == 0]
data_1=df.loc[df['label'] == 1]
print(data_1.shape)
print(data_0.shape)
print(df.shape)

(48, 3)
(32, 3)
(80, 3)


In [5]:
data_mean_0=data_0.groupby('label').mean()
data_mean_1=data_1.groupby('label').mean()
data_var_0=data_0.groupby('label').var()
data_var_1=data_1.groupby('label').var()
co_var_0=np.cov(data_0['x'],data_0['y'])
siginv_0=np.linalg.inv(co_var_0)
co_var_1=np.cov(data_1['x'],data_1['y'])
siginv_1=np.linalg.inv(co_var_1)
data_mean_1

x          y
label                      
1      75.518831  73.855421

In [6]:
x_mean_0=data_mean_0['x'].values[0]
y_mean_0=data_mean_0['y'].values[0]
x_mean_1=data_mean_1['x'].values[0]
y_mean_1=data_mean_1['y'].values[0]
print(x_mean_0)

51.78408060051157


In [7]:
def p_a_given_b(a,siginv,mean,co_var,p1):
    detm=np.linalg.det(co_var)
    mat=a - mean
    p=1/(np.sqrt(2*np.pi*detm))*np.exp((-(np.dot(np.dot(mat,siginv),mat.T)))/2)
    #print("mat=",mat,"np=",np.dot(np.dot(mat,siginv),mat.T),"exp=",np.exp((-(np.dot(np.dot(mat,siginv),mat.T)))/2))
    #print("deno=",(np.sqrt(2*np.pi*detm)),"detm=",detm)
    return p1*p

In [8]:
def p_c_given_d(a,mean1,co_var1,p1):
    for x,mean,co_var in zip(a,mean1,co_var1):
        p=1
        #print("a=",x,"mean=",mean,"cov=",co_var)
        p*=1/(np.sqrt(2*np.pi*co_var))*np.exp((-(x-mean)**2)/(2*co_var))
        #print("p=",p,"deno=",(np.sqrt(2*np.pi*co_var)),"exp=",np.exp((-(x-mean)**2)/(2*co_var)))
        #print("pow=")
        return p1*p
    

In [9]:
l=np.array(df2.loc[:,:'y'].values[18])
zero=p_c_given_d(l,np.array(data_mean_0.values[0]),np.array(data_var_0.values[0]),32/80)
one=p_c_given_d(l,np.array(data_mean_1.values[0]),np.array(data_mean_1.values[0]),48/80)
l0=zero/(zero+one)
l1=one/(zero+one)
print("l0=",l0,"l1=",l1)
if l0>0.5 and l0>l1:
    print("clasa=0")
else:
    print("class=1")

l0= 0.17470924150986 l1= 0.82529075849014
class=1


In [10]:
list1=[]
list2=[]
l=np.array(df2.loc[:,:'y'].values[4])
for i in range(0,20):
    l=np.array(df2.loc[:,:'y'].values[i])
    zero=p_a_given_b(l,siginv_0,np.array(data_mean_0.values[0]),co_var_0,32/80)
    one=p_a_given_b(l,siginv_1,np.array(data_mean_1.values[0]),co_var_1,48/80)
    l0=zero/(zero+one)
    l1=one/(zero+one)
    #print("l0=",l0,"l1=",l1)
    if l0>0.5 and l0>l1:
        #print("clasa=0")
        list1.append(0)
    else:
        list1.append(1)
        #print("class=1")
print(list1)

[0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1]


In [11]:
for i in (df2.loc[:,'label':]).values:
    list2.append(i[0])
print(list2)

[0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1]


In [12]:
count=0
miscount=0
for i,j in zip(list1,list2):
    if i==j:
        count+=1
print("acurracy=",(count/len(list1))*100)

acurracy= 85.0


Got acurracy =85% due to small dataset and having small no. of features (i.e only two).This dataset contain 100 example that are splited into 5:1 ratio of training and testing part.